src/libasr/pass/intrinsic_array_function_registry.h

```diff
@@ -1316,6 +1318,265 @@ namespace Cshift {
+        int array_dim = -1;
+        extract_value(array_dims[0].m_length, array_dim);
+        ASRUtils::require_impl(array_rank > 0, "The argument `array` in `eoshift` must be of rank > 0", array->base.loc, diag);
+        ASRBuilder b(al, loc);
+        Vec<ASR::dimension_t> result_dims; result_dims.reserve(al, 1);
+        int overload_id = 2;
+        if (!is_dim_present) {
+            result_dims.push_back(al, b.set_dim(array_dims[0].m_start, array_dims[0].m_length));
+            ret_type = ASRUtils::duplicate_type(al, ret_type, &result_dims);
+        } else {
+            result_dims.push_back(al, b.set_dim(dim, dim));
+            ret_type = ASRUtils::duplicate_type(al, ret_type, &result_dims);
+        }
+        if (is_type_allocatable) {
+            ret_type = TYPE(ASR::make_Allocatable_t(al, loc, ret_type));
+        }
+        ASR::expr_t *final_boundary = nullptr;
+        if(is_boundary_present){
+            final_boundary = boundary;
+        }
+        else{
+            if(is_integer(*type_array))
+                final_boundary = b.i_t(0, int32);
+            else if(is_real(*type_array))
+                final_boundary = b.f_t(0.0, real32);
+            else if(is_logical(*type_array))
+                final_boundary = b.bool_t(false, logical);
+            else if(is_character(*type_array))
+                final_boundary = b.StringConstant("", type_array);
+        }
+        Vec<ASR::expr_t*> m_args; m_args.reserve(al, 3);
+        m_args.push_back(al, array); m_args.push_back(al, shift); m_args.push_back(al, final_boundary);
+        ASR::expr_t *value = nullptr;
+        if (all_args_evaluated(m_args)) {
+            value = eval_Eoshift(al, loc, ret_type, m_args, diag);
+        }
+        return make_IntrinsicArrayFunction_t_util(al, loc,
+            static_cast<int64_t>(IntrinsicArrayFunctions::Eoshift),
+            m_args.p, m_args.n, overload_id, ret_type, value);
+    }
+
+    static inline ASR::expr_t* instantiate_Eoshift(Allocator &al,
+            const Location &loc, SymbolTable *scope,
+            Vec<ASR::ttype_t*> &arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t> &m_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_eoshift");
+        fill_func_arg("array", duplicate_type_with_empty_dims(al, arg_types[0]));
+        fill_func_arg("shift", arg_types[1]);
+        fill_func_arg("boundary", arg_types[2]);
+        ASR::ttype_t* return_type_ = return_type;
+        /*
+            Eoshift(array, shift, dim)
+            int i = 0
+            for j = shift, size(array)
+                result[i] = array[j]
+                i = i + 1
+            end for
+            for j = 1, shift
+                result[i] = array[j]
+                i = i + 1
+            end for
+
+            if (k shift= 0) then
+                i = size(array)
+                i = i - shift
+                i = i + 1
```

Comment:
```suggestion
                i = size(array) - shift + 1
```
Can this be written as this?